### PR TITLE
feat: Listing the policies applied on containers in Un-orchestrated environment

### LINKED ISSUE #1311: list the applied policies for unorchestrated workloads
We need to list the applied policies on unorchestrated workloads using `karmor probe`, the same way we list the applied policies for orchestrted workloads.

**Describe the solution you'd like**
we can add grpc endpoint and use it to list the policies for workloads.

### PR DESCRIPTION
**Pull request by Aryan-sharma11**
**Purpose of PR?**:

Fixes #1311 
image
image
This PR exposes the Policy and containers List which will  be utilised by Kubearmor client in systemd mode. 

**Does this PR introduce a breaking change?** NA

**If the changes in this PR are manually verified, list down the scenarios covered:**:

**Additional information for reviewer?** :
_Mention if this PR is part of any design or a continuation of previous PRs_

**Checklist:**
- [ ] Bug fix. Fixes #<issue number>
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] PR Title follows the convention of  `<type>(<scope>): <subject>`
- [x] Commit has unit tests
- [x] Commit has integration tests

### CODE REVIEW

File: protobuf/karmor.proto [hunk 1]
```diff
@@ -0,0 +1,21 @@
+syntax = "proto3";
+import "google/protobuf/empty.proto";
+package karmor;
+
+option go_package="github.com/kubearmor/KubeArmor/protobuf";
+
+message ContainerData {
```
Comment (daemon1024) [inline protobuf/karmor.proto]: Why is this a new proto? Shouldn't it be part of the Policy Proto? WDYT?

**Review by daemon1024 — CHANGES_REQUESTED** [review]
Did a quick review, one review in line. Can you handle lint errors as well?

**Comment by ShubhamTatvamasi** [comment]
@Aryan-sharma11 can you please rebase and run the CI tests again?

**Review by daemon1024 — APPROVED** [review]

File: KubeArmor/core/karmorprobedata.go [hunk 2]
```diff
@@ -62,3 +72,77 @@ func (dm *KubeArmorDaemon) SetKarmorData() {
...
+	// Mapping Hostpolicies to their host hostName : HostPolicy
+	hostMap := make(map[string]*pb.HostSecurityPolicies)
+
+	dm.HostSecurityPoliciesLock.Lock()
+	for _, hp := range dm.HostSecurityPolicies {
+
+		hostName := ""
+		for _, v := range hp.Spec.NodeSelector.MatchLabels {
+			hostName = v
+		}
```
Comment (DelusionalOptimist) [inline KubeArmor/core/karmorprobedata.go]: Are we looking for a specific label to match here? Why do we need this loop?
Also, KubeArmor only protects a single node anyways. The name is stored in `dm.Node.NodeName`. Shouldn't we directly use that?

File: KubeArmor/core/karmorprobedata.go [hunk 3]
```diff
@@ -62,3 +72,77 @@ func (dm *KubeArmorDaemon) SetKarmorData() {
...
+	dm.HostSecurityPoliciesLock.Unlock()
+
+	return containerlist, containerMap, hostMap
+
+}
+
+// GetProbeData() sends policy data through grpc client
+func (p *Probe) GetProbeData(c context.Context, in *empty.Empty) (*pb.ProbeResponse, error) {
+
+	var containerMap map[string]*pb.ContainerData
```
Comment (DelusionalOptimist) [inline KubeArmor/core/karmorprobedata.go]: This declaration is not needed.
```suggestion
```

File: protobuf/go.mod [hunk 4]
```diff
@@ -11,12 +11,12 @@ replace (
 )
 
 require (
+	github.com/golang/protobuf v1.5.3
```
Comment (DelusionalOptimist) [inline protobuf/go.mod]: Can we stick to `google.golang.org/protobuf`? Mixing protobuf dependecies has led to terrible runtime panics in the past : (

File: .github/workflows/ci-test-ginkgo.yml
Comment (DelusionalOptimist) [inline .github/workflows/ci-test-ginkgo.yml:1]: Also need to update other workflows where ginkgo runs. For instance `ci-latest-release.yaml`, `ci-test-controllers`.

File: tests/util/kartutil.go [hunk 5]
```diff
@@ -598,3 +602,44 @@ func K8sRuntimeEnforcer() string {
 	runtimeEnforcer := nodes.Items[0].Labels["kubearmor.io/enforcer"]
 	return runtimeEnforcer
 }
+
+// CommandParse executes docker commmands
+func CommandParse(cmdstr string) (string, error) {
```
Comment (DelusionalOptimist) [inline tests/util/kartutil.go]:
```suggestion
func RunDockerCommand(cmdstr string) (string, error) {
```

**Review by DelusionalOptimist — CHANGES_REQUESTED** [review]
@Aryan-sharma11 Mostly looking good. Please find some comments inline.

Though I couldn't tryout probe with kubearmor-client, I did manage to get the probe output with grpcurl.


image

In the above, I created a policy for ubuntu-vm but then deleted it. However it still showed up in the response.
Also, ubuntu-3 container didn't exist but it still showed up in the "containerMap" as it had a policy. This is fine but IMO it shouldn't have policyEnabled = 1 as this container doesn't exist. Though I think this information is taken directly from stated stored in KubeArmor but we might need to handle these edge cases either here or in client.

**Comment by Aryan-sharma11** [comment]
> @Aryan-sharma11 Mostly looking good. Please find some comments inline.
> 
> Though I couldn't tryout probe with kubearmor-client, I did manage to get the probe output with grpcurl.
> 
> image
> 
> In the above, I created a policy for ubuntu-vm but then deleted it. However it still showed up in the response. Also, ubuntu-3 container didn't exist but it still showed up in the "containerMap" as it had a policy. This is fine but IMO it shouldn't have policyEnabled = 1 as this container doesn't exist. Though I think this information is taken directly from stated stored in KubeArmor but we might need to handle these edge cases either here or in client.



> @Aryan-sharma11 Mostly looking good. Please find some comments inline.
> 
> Though I couldn't tryout probe with kubearmor-client, I did manage to get the probe output with grpcurl.
> 
> image
> 
> In the above, I created a policy for ubuntu-vm but then deleted it. However it still showed up in the response. Also, ubuntu-3 container didn't exist but it still showed up in the "containerMap" as it had a policy. This is fine but IMO it shouldn't have policyEnabled = 1 as this container doesn't exist. Though I think this information is taken directly from stated stored in KubeArmor but we might need to handle these edge cases either here or in client.

Yes In the client we handle this case of policy being there and container not being present as we match the container Map with the Container List so only the active containers are shown.  for [ref](https://github.com/kubearmor/kubearmor-client/pull/372/files#diff-0b3ca41e370fff5ec9331a998f39c4c430b7bac2ceb0318fcb2a98f31e178b03R631).

**Review by DelusionalOptimist — APPROVED** [review]
LGTM. :rocket:

**Pull request merged**